# Data Understanding & Initial Investigation

## Objective

This notebook investigates the structure, grain, entities, data types, and initial data quality of the DataCo Smart Supply Chain dataset.

The objective at this stage is **understanding before cleaning**.

We will answer:

1. How large is the dataset?
2. What does each row represent?
3. What entities and identifiers exist?
4. What data types are present?
5. Are there missing or duplicate records?
6. What date and time fields are available?
7. What potential data-quality issues need to be addressed?

In [1]:
import pandas as pd
import numpy as np

print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)

Pandas version: 3.0.5
NumPy version: 2.5.2


In [3]:
# Loading the main DataCo supply chain dataset

file_path = "../data/raw/DataCoSupplyChainDataset.csv"

df = pd.read_csv(
    file_path,
    encoding="latin-1"
)

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [4]:
# Dataset dimensions

rows, columns = df.shape

print(f"Rows: {rows:,}")
print(f"Columns: {columns:,}")

Rows: 180,519
Columns: 53


In [5]:
# List every column in the dataset

for i, column in enumerate(df.columns, start=1):
    print(f"{i:2}. {column}")

 1. Type
 2. Days for shipping (real)
 3. Days for shipment (scheduled)
 4. Benefit per order
 5. Sales per customer
 6. Delivery Status
 7. Late_delivery_risk
 8. Category Id
 9. Category Name
10. Customer City
11. Customer Country
12. Customer Email
13. Customer Fname
14. Customer Id
15. Customer Lname
16. Customer Password
17. Customer Segment
18. Customer State
19. Customer Street
20. Customer Zipcode
21. Department Id
22. Department Name
23. Latitude
24. Longitude
25. Market
26. Order City
27. Order Country
28. Order Customer Id
29. order date (DateOrders)
30. Order Id
31. Order Item Cardprod Id
32. Order Item Discount
33. Order Item Discount Rate
34. Order Item Id
35. Order Item Product Price
36. Order Item Profit Ratio
37. Order Item Quantity
38. Sales
39. Order Item Total
40. Order Profit Per Order
41. Order Region
42. Order State
43. Order Status
44. Order Zipcode
45. Product Card Id
46. Product Category Id
47. Product Description
48. Product Image
49. Product Name
50. Product

In [8]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [9]:
pd.set_option("display.max_columns", None)

df.head()


,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [10]:
# Investigate identifier uniqueness

print("Total rows:", len(df))
print("Unique Order IDs:", df["Order Id"].nunique())
print("Unique Order Item IDs:", df["Order Item Id"].nunique())
print("Unique Customer IDs:", df["Customer Id"].nunique())
print("Unique Product IDs:", df["Product Card Id"].nunique())

Total rows: 180519
Unique Order IDs: 65752
Unique Order Item IDs: 180519
Unique Customer IDs: 20652
Unique Product IDs: 118


In [11]:
# Investigate order-item structure

order_item_counts = df.groupby("Order Id")["Order Item Id"].nunique()

print(order_item_counts.describe())

count    65752.000000
mean         2.745453
std          1.478363
min          1.000000
25%          1.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: Order Item Id, dtype: float64


In [12]:
# Distribution of items per order

print("Orders with 1 item:", (order_item_counts == 1).sum())
print("Orders with 2 items:", (order_item_counts == 2).sum())
print("Orders with 3 items:", (order_item_counts == 3).sum())
print("Orders with 4 items:", (order_item_counts == 4).sum())
print("Orders with 5 items:", (order_item_counts == 5).sum())

Orders with 1 item: 19850
Orders with 2 items: 11447
Orders with 3 items: 11398
Orders with 4 items: 11704
Orders with 5 items: 11353


In [13]:
# Missing-value assessment

missing_values = df.isnull().sum()

missing_values = missing_values[missing_values > 0].sort_values(ascending=False)

print("Columns with missing values:")
print(missing_values)

Columns with missing values:
Product Description    180519
Order Zipcode          155679
Customer Lname              8
Customer Zipcode            3
dtype: int64


In [14]:
# Missing-value percentage

missing_percentage = (
    df.isnull().mean() * 100
).sort_values(ascending=False)

missing_percentage = missing_percentage[missing_percentage > 0]

print("Missing-value percentage:")
print(missing_percentage)

Missing-value percentage:
Product Description    100.000000
Order Zipcode           86.239676
Customer Lname           0.004432
Customer Zipcode         0.001662
dtype: float64


### Missing-Value Findings

The main dataset contains 180,519 order-item records across 53 columns.

The missing-value assessment identified four columns with missing values:

- `Product Description` is completely missing across all records and therefore contains no usable information.
- `Order Zipcode` is missing for approximately 86.24% of records and requires further assessment before deciding whether to retain it.
- `Customer Lname` has only 8 missing values.
- `Customer Zipcode` has only 3 missing values.

No missing values are currently being modified or imputed. Cleaning decisions will be made after the complete data-quality assessment.

In [15]:
# Duplicate record assessment

duplicate_rows = df.duplicated().sum()

print("Duplicate rows:", duplicate_rows)
print("Duplicate percentage:", (duplicate_rows / len(df)) * 100)

Duplicate rows: 0
Duplicate percentage: 0.0


In [16]:
# Duplicate Order Item IDs

duplicate_order_items = df["Order Item Id"].duplicated().sum()

print("Duplicate Order Item IDs:", duplicate_order_items)

Duplicate Order Item IDs: 0


### Duplicate-Record Findings

The duplicate assessment found no completely duplicated records in the main dataset.

- Total records: 180,519
- Duplicate rows: 0
- Duplicate `Order Item Id` values: 0

The `Order Item Id` therefore provides a unique identifier for each order-item record in the dataset. Combined with the earlier order-level analysis, this supports the conclusion that the dataset is structured at the order-item/line-item grain.

In [17]:
print(df.dtypes)

Type                                 str
Days for shipping (real)           int64
Days for shipment (scheduled)      int64
Benefit per order                float64
Sales per customer               float64
Delivery Status                      str
Late_delivery_risk                 int64
Category Id                        int64
Category Name                        str
Customer City                        str
Customer Country                     str
Customer Email                       str
Customer Fname                       str
Customer Id                        int64
Customer Lname                       str
Customer Password                    str
Customer Segment                     str
Customer State                       str
Customer Street                      str
Customer Zipcode                 float64
Department Id                      int64
Department Name                      str
Latitude                         float64
Longitude                        float64
Market          

In [18]:


print(df.dtypes.value_counts())

str        24
float64    15
int64      14
Name: count, dtype: int64


In [19]:
# Identify date-related columns

date_columns = [
    col for col in df.columns
    if "date" in col.lower()
]

print("Date-related columns:")
print(date_columns)

print("\nTheir data types:")
print(df[date_columns].dtypes)

Date-related columns:
['order date (DateOrders)', 'shipping date (DateOrders)']

Their data types:
order date (DateOrders)       str
shipping date (DateOrders)    str
dtype: object


### Data Type Findings

The raw dataset contains 53 columns with the following data types:

- 24 string columns
- 14 integer columns
- 15 floating-point columns

Two important date fields are currently stored as strings:

- `order date (DateOrders)`
- `shipping date (DateOrders)`

These will be converted to proper datetime types during the data-cleaning stage.

Several numeric columns require additional consideration. In particular, identifier fields such as `Customer Id`, `Order Id`, and `Product Card Id` are stored as integers but represent identifiers rather than continuous numerical measurements.

`Customer Zipcode` and `Order Zipcode` are stored as floating-point values because they contain missing values and therefore require careful handling.

`Product Description` is stored as `float64` and was found to be completely missing. This column will be evaluated for removal during the cleaning stage.

At this stage, no transformations have been applied because the purpose of this notebook is to understand and document the raw dataset before cleaning.

In [20]:
# Inspect the date range of the raw dataset

order_dates = pd.to_datetime(
    df["order date (DateOrders)"],
    errors="coerce"
)

shipping_dates = pd.to_datetime(
    df["shipping date (DateOrders)"],
    errors="coerce"
)

print("Order date range:")
print("Earliest:", order_dates.min())
print("Latest:", order_dates.max())

print("\nShipping date range:")
print("Earliest:", shipping_dates.min())
print("Latest:", shipping_dates.max())

Order date range:
Earliest: 2015-01-01 00:00:00
Latest: 2018-01-31 23:38:00

Shipping date range:
Earliest: 2015-01-03 00:00:00
Latest: 2018-02-06 22:14:00


In [21]:
print("Invalid order dates:", order_dates.isna().sum())
print("Invalid shipping dates:", shipping_dates.isna().sum())

Invalid order dates: 0
Invalid shipping dates: 0


In [22]:
invalid_date_sequence = (shipping_dates < order_dates).sum()

print("Shipping dates before order dates:", invalid_date_sequence)

Shipping dates before order dates: 0


### Date Range & Temporal Integrity Findings

The dataset contains approximately three years of historical order activity.

- **Order date range:** 2015-01-01 to 2018-01-31
- **Shipping date range:** 2015-01-03 to 2018-02-06
- **Invalid order dates:** 0
- **Invalid shipping dates:** 0
- **Shipping dates occurring before order dates:** 0

The order and shipping date fields can be successfully parsed without invalid values, and no records were found where the shipping date precedes the order date.

The shipping period extends slightly beyond the order period, which is expected because orders placed near the end of the dataset may have been shipped after the final order date.

The approximately three-year order history provides a suitable temporal foundation for later demand trend, seasonality, and forecasting analysis.

The date columns will be converted from strings to proper datetime types during the data-cleaning stage.

In [23]:
# Investigate important categorical variables

categorical_columns = [
    "Type",
    "Delivery Status",
    "Customer Segment",
    "Market",
    "Department Name",
    "Shipping Mode",
    "Order Status"
]

for column in categorical_columns:
    print(f"\n{'=' * 60}")
    print(column)
    print(f"{'=' * 60}")
    print(df[column].value_counts(dropna=False))
    


Type
Type
DEBIT       69295
TRANSFER    49883
PAYMENT     41725
CASH        19616
Name: count, dtype: int64

Delivery Status
Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

Customer Segment
Customer Segment
Consumer       93504
Corporate      54789
Home Office    32226
Name: count, dtype: int64

Market
Market
LATAM           51594
Europe          50252
Pacific Asia    41260
USCA            25799
Africa          11614
Name: count, dtype: int64

Department Name
Department Name
Fan Shop              66861
Apparel               48998
Golf                  33220
Footwear              14525
Outdoors               9686
Fitness                2479
Discs Shop             2026
Technology             1465
Pet Shop                492
Book Shop               405
Health and Beauty       362
Name: count, dtype: int64

Shipping Mode
Shipping Mode
Standard Class    107752
Second Class       35216
Fir

In [24]:
# Investigate geographic dimensions

geographic_columns = [
    "Customer Country",
    "Order Country",
    "Order Region"
]

for column in geographic_columns:
    print(f"\n{column}")
    print("-" * 40)
    print("Unique values:", df[column].nunique())
    print(df[column].value_counts().head(15))


Customer Country
----------------------------------------
Unique values: 2
Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

Order Country
----------------------------------------
Unique values: 164
Order Country
Estados Unidos          24840
Francia                 13222
México                  13172
Alemania                 9564
Australia                8497
Brasil                   7987
Reino Unido              7302
China                    5758
Italia                   4989
India                    4783
Indonesia                4204
España                   3868
El Salvador              3726
República Dominicana     3669
Honduras                 3629
Name: count, dtype: int64

Order Region
----------------------------------------
Unique values: 23
Order Region
Central America    28341
Western Europe     27109
South America      14935
Oceania            10148
Northern Europe     9792
Southeast Asia      9539
Southern Europe     9431
Caribbean  

In [25]:
# Investigate product and category dimensions

product_columns = [
    "Category Name",
    "Product Name"
]

for column in product_columns:
    print(f"\n{column}")
    print("-" * 40)
    print("Unique values:", df[column].nunique())
    print(df[column].value_counts().head(15))


Category Name
----------------------------------------
Unique values: 50
Category Name
Cleats                  24551
Men's Footwear          22246
Women's Apparel         21035
Indoor/Outdoor Games    19298
Fishing                 17325
Water Sports            15540
Camping & Hiking        13729
Cardio Equipment        12487
Shop By Sport           10984
Electronics              3156
Accessories              1780
Golf Balls               1475
Girls' Apparel           1201
Golf Gloves              1070
Trade-In                  974
Name: count, dtype: int64

Product Name
----------------------------------------
Unique values: 118
Product Name
Perfect Fitness Perfect Rip Deck                 24515
Nike Men's CJ Elite 2 TD Football Cleat          22246
Nike Men's Dri-FIT Victory Golf Polo             21035
O'Brien Men's Neoprene Life Vest                 19298
Field & Stream Sportsman 16 Gun Fire Safe        17325
Pelican Sunstream 100 Kayak                      15500
Diamondback Women's

### Categorical & Business Dimension Findings

The categorical investigation reveals several important dimensions of the supply chain dataset.

#### Transaction Types

Four transaction types are present:

- DEBIT
- TRANSFER
- PAYMENT
- CASH

These may be useful for later analysis of payment behavior and its relationship with order status and profitability.

#### Delivery Status

Four delivery outcomes are present:

- Late delivery
- Advance shipping
- Shipping on time
- Shipping canceled

Late delivery represents the largest category by record count. Because the dataset is stored at the order-item level, these counts should not be interpreted directly as unique orders or customers. Delivery performance will therefore be analyzed at the appropriate grain during later stages.

#### Customer Segments

Three customer segments are present:

- Consumer
- Corporate
- Home Office

Consumer is the largest segment by record count. Later analysis will compare these segments using business metrics such as revenue, profitability, order behavior, and delivery performance rather than record counts alone.

#### Geographic Dimensions

The dataset contains two distinct geographic perspectives:

- Customer Country: 2 unique values
- Order Country: 164 unique values
- Order Region: 23 unique values

The difference between customer location and order destination indicates that these fields represent different business dimensions and should be retained separately.

#### Product Dimensions

The dataset contains:

- 50 product categories
- 118 unique products

This provides sufficient product-level granularity for later revenue, profitability, demand, and product-performance analysis.

Overall, the categorical investigation confirms that the dataset contains multiple business dimensions spanning customers, transactions, products, geography, logistics, and order operations.

In [26]:
# Inspect numerical variables

numeric_columns = df.select_dtypes(
    include=["int64", "float64"]
).columns

print("Number of numerical columns:", len(numeric_columns))
print("\nNumerical columns:")
for column in numeric_columns:
    print("-", column)

Number of numerical columns: 29

Numerical columns:
- Days for shipping (real)
- Days for shipment (scheduled)
- Benefit per order
- Sales per customer
- Late_delivery_risk
- Category Id
- Customer Id
- Customer Zipcode
- Department Id
- Latitude
- Longitude
- Order Customer Id
- Order Id
- Order Item Cardprod Id
- Order Item Discount
- Order Item Discount Rate
- Order Item Id
- Order Item Product Price
- Order Item Profit Ratio
- Order Item Quantity
- Sales
- Order Item Total
- Order Profit Per Order
- Order Zipcode
- Product Card Id
- Product Category Id
- Product Description
- Product Price
- Product Status


In [27]:
# Generate descriptive statistics for numerical variables
df[numeric_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.497654,1.623722,0.000000,2.000000,3.000000,5.000000,6.000000
Days for shipment (scheduled),180519.0,2.931847,1.374449,0.000000,2.000000,4.000000,4.000000,4.000000
Benefit per order,180519.0,21.974989,104.433526,-4274.979980,7.000000,31.520000,64.800003,911.799988
Sales per customer,180519.0,183.107609,120.043670,7.490000,104.379997,163.990005,247.399994,1939.989990
Late_delivery_risk,180519.0,0.548291,0.497664,0.000000,0.000000,1.000000,1.000000,1.000000
Category Id,180519.0,31.851451,15.640064,2.000000,18.000000,29.000000,45.000000,76.000000
Customer Id,180519.0,6691.379495,4162.918106,1.000000,3258.500000,6457.000000,9779.000000,20757.000000
Customer Zipcode,180516.0,35921.126914,37542.461122,603.000000,725.000000,19380.000000,78207.000000,99205.000000
Department Id,180519.0,5.443460,1.629246,2.000000,4.000000,5.000000,7.000000,12.000000
Latitude,180519.0,29.719955,9.813646,-33.937553,18.265432,33.144863,39.279617,48.781933


In [28]:
# Check minimum values of numerical columns

numeric_min = df[numeric_columns].min().sort_values()

print("Minimum values:")
print(numeric_min)

Minimum values:
Benefit per order               -4274.979980
Order Profit Per Order          -4274.979980
Longitude                        -158.025986
Latitude                          -33.937553
Order Item Profit Ratio            -2.750000
Order Item Discount                 0.000000
Order Item Discount Rate            0.000000
Days for shipping (real)            0.000000
Days for shipment (scheduled)       0.000000
Product Status                      0.000000
Late_delivery_risk                  0.000000
Order Id                            1.000000
Order Item Quantity                 1.000000
Order Customer Id                   1.000000
Customer Id                         1.000000
Order Item Id                       1.000000
Category Id                         2.000000
Product Category Id                 2.000000
Department Id                       2.000000
Sales per customer                  7.490000
Order Item Total                    7.490000
Sales                               9.9

### Numerical Variable Findings

The numerical investigation identified 29 numerical columns. These include identifiers, business measures, operational metrics, and geographic variables.

Several important observations were identified:

- Identifier columns such as `Order Id`, `Customer Id`, `Product Card Id`, and `Order Item Id` are numeric but represent entities rather than continuous measurements.
- `Order Item Quantity` ranges from 1 to 5, with a median of 1.
- `Product Price` and `Order Item Product Price` range from approximately 9.99 to 1999.99.
- `Late_delivery_risk` is a binary variable containing values from 0 to 1.
- `Days for shipping (real)` ranges from 0 to 6 days.
- `Days for shipment (scheduled)` ranges from 0 to 4 days.
- Profit-related measures contain both positive and negative values. Negative profit observations will not be treated as errors without further investigation because they may represent genuinely unprofitable transactions.
- `Order Zipcode` has a high proportion of missing values and will require careful treatment.
- `Product Description` contains no non-null observations and is therefore a strong candidate for removal during the cleaning stage.
- No conclusions about numerical outliers or invalid business values will be made solely from descriptive statistics. Potential anomalies will be investigated using business rules and cross-column relationships before cleaning.

The numerical investigation therefore identifies several areas requiring validation rather than immediate correction.

In [29]:
# Validate relationships between important numerical fields

print("Sales - Discount = Order Item Total:")

total_difference = (
    df["Sales"] - df["Order Item Discount"] - df["Order Item Total"]
).abs()

print("Maximum absolute difference:", total_difference.max())
print("Rows with difference > 0.01:", (total_difference > 0.01).sum())


print("\nProduct Price vs Order Item Product Price:")

price_difference = (
    df["Product Price"] - df["Order Item Product Price"]
).abs()

print("Maximum absolute difference:", price_difference.max())
print("Rows with difference > 0.01:", (price_difference > 0.01).sum())

Sales - Discount = Order Item Total:
Maximum absolute difference: 0.010013549999996485
Rows with difference > 0.01: 1224

Product Price vs Order Item Product Price:
Maximum absolute difference: 0.0
Rows with difference > 0.01: 0


In [30]:
# Validate quantity and delivery-risk ranges

print("Unique Order Item Quantity values:")
print(sorted(df["Order Item Quantity"].unique()))

print("\nUnique Late Delivery Risk values:")
print(sorted(df["Late_delivery_risk"].unique()))

print("\nUnique Product Status values:")
print(sorted(df["Product Status"].unique()))

Unique Order Item Quantity values:
[np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)]

Unique Late Delivery Risk values:
[np.int64(0), np.int64(1)]

Unique Product Status values:
[np.int64(0)]


In [31]:
# Check whether Sales is consistent with price × quantity

sales_difference = (
    df["Order Item Product Price"] * df["Order Item Quantity"]
    - df["Sales"]
).abs()

print("Maximum absolute difference between price × quantity and Sales:",
      sales_difference.max())

print("Rows with difference > 0.01:",
      (sales_difference > 0.01).sum())

Maximum absolute difference between price × quantity and Sales: 2.289999997628911e-05
Rows with difference > 0.01: 0


In [32]:
# Validate Sales - Discount = Order Item Total after rounding

total_difference_rounded = (
    (
        df["Sales"]
        - df["Order Item Discount"]
        - df["Order Item Total"]
    )
    .round(2)
    .abs()
)

print("Maximum difference after rounding:",
      total_difference_rounded.max())

print("Rows with difference > 0.01 after rounding:",
      (total_difference_rounded > 0.01).sum())

Maximum difference after rounding: 0.01
Rows with difference > 0.01 after rounding: 0


## Data Understanding Summary

The initial investigation established the structure, quality, and major business dimensions of the raw DataCo supply chain dataset.

### Dataset Structure

- 180,519 order-item records
- 53 columns
- 65,752 unique orders
- 180,519 unique order items
- 20,652 unique customers
- 118 unique products
- 50 product categories

The dataset is primarily at the **order-item level**, meaning a single order may contain multiple records. This distinction will be maintained throughout the analysis to avoid confusing order-item counts with unique order counts.

### Data Quality

- No duplicate rows were identified.
- No duplicate `Order Item Id` values were identified.
- Order and shipping dates contain no invalid values.
- No records were found where the shipping date occurs before the order date.
- `Product Description` is completely missing.
- `Order Zipcode` contains a substantial proportion of missing values.
- `Customer Lname` and `Customer Zipcode` contain only a very small number of missing values.

### Data Type Observations

The raw dataset contains 24 string columns, 14 integer columns, and 15 floating-point columns.

The order and shipping date fields are currently stored as strings and will be converted to datetime types during the cleaning stage.

Several numeric fields represent identifiers rather than continuous measurements and will be treated accordingly during analysis.

### Business Dimensions

The dataset contains multiple analytical dimensions covering:

- Customers and customer segments
- Products and product categories
- Departments
- Markets and geographic regions
- Orders and order items
- Shipping modes
- Delivery performance
- Transaction types
- Revenue and profitability

### Numerical Validation

Several internal consistency checks were performed:

- `Product Price` and `Order Item Product Price` are identical across all records.
- `Sales` is consistent with `Order Item Product Price × Order Item Quantity` within floating-point precision.
- `Order Item Total` is consistent with `Sales − Order Item Discount` within a one-cent tolerance.
- `Order Item Quantity` contains only values from 1 to 5.
- `Late_delivery_risk` contains only binary values (0 and 1).
- `Product Status` contains only the value 0 and therefore has no variation.

### Important Analytical Considerations

Negative profit values are present in the dataset. These observations will not be removed automatically because negative profitability may represent genuine business outcomes rather than data errors.

Similarly, extreme values will be investigated using business rules and cross-column relationships before any records are removed or modified.

The investigation therefore provides a documented baseline of the raw dataset before transformation.

### Next Phase

The next stage is to build a **reproducible data-cleaning and transformation pipeline**.

The raw dataset will remain unchanged. Cleaning and feature engineering will be performed on a separate analytical dataframe so that the original source data remains reproducible.